### ISSUE TITLE
winlogbeat setup should succeed on non-windows machines.

### ISSUE THREAD

**Issue by jordansissel** [elastic/beats#13707]
Two main problems:

1) No linux releases are available with winlogbeat.
2) `winlogbeat setup` on Linux fails.

item 1 can be worked around (for now, not sure if this is by accident), by building myself:

```
% git checkout v7.3.1
% make
go build -i -ldflags "-X github.com/elastic/beats/libbeat/version.buildTime=2019-09-16T23:51:00Z -X github.com/elastic/beats/libbeat/version.commit=a4be71b90ce3e3b8213b616adfcd9e455513da45"
% ./winlogbeat setup
Exiting: Failed to create new event log. No event log API is available on this system
```

This fails in beater/winlogbeat.go's `init`  call because Windows Event log API is attempted even when only doing `setup` step.

**Comment by jordansissel** [comment]
I'm attempting a workaround, but got stuck in some complexities of beats build system. The reason I'm trying to workaround this is two fold:

1) I have a project requiring winlogbeat that's due in ~2 weeks, blocked on needing `winlogbeat setup`
2) I don't have _any_ windows machines in production. 100% of my systems are on Linux in Kubernetes, so adding a one-off Windows VM is going to be very weird.
3) Automation of setup is greatly preferable as we have dozens of beat product and version combinations. Running this manually on a Windows VM is going to be too painful, I think, especially as the Elastic Stack averages two releases a month.

First for this workaround, we need to remove the `init`:

```
% sed -i -e 's/^\(func .* \)\(init(.*) error {.*\)$/\1\2 return nil }\n\1_\2/' beater/winlogbeat.go
```

Rebuilding, this gets me further, but kibana's objects are missing:

```
% ./winlogbeat setup
Index setup finished.
Loading dashboards (Kibana must be running and reachable)
Skipping loading dashboards, No directory /home/jls/projects/go/src/github.com/elastic/beats/winlogbeat/kibana/7
```

Attempting to generate this stuff with `make update` brings me along my yak shave adventure (hehe) to missing `mage` and then missing `virtualenv`:

```
% make update
Installing mage v1.8.0 from vendor dir.
/home/jls/.magefile cleaned
mage update
2019/09/16 16:55:44 Found Elastic Beats dir at /home/jls/projects/go/src/github.com/elastic/beats
>> update:moduleDocs: Collecting module documentation for security, sysmon.
>> Building winlogbeat.yml for windows/amd64
>> Building winlogbeat.reference.yml for windows/amd64
Generated fields.yml for winlogbeat to /home/jls/projects/go/src/github.com/elastic/beats/winlogbeat/build/fields/fields.yml
Generated fields.yml for winlogbeat to /home/jls/projects/go/src/github.com/elastic/beats/winlogbeat/build/fields/fields.all.yml
Generated fields.yml for winlogbeat to /home/jls/projects/go/src/github.com/elastic/beats/winlogbeat/build/fields/fields.common.yml
Error: failed to run "virtualenv /home/jls/projects/go/src/github.com/elastic/beats/build/ve/linux: exec: "virtualenv": executable file not found in $PATH"
Makefile:12: recipe for target 'update' failed
make: *** [update] Error 1
```

At this point, I've stopped trying to workaround this for today.

**Comment by jordansissel** [comment]
Finished the workaround. This Dockerfile works for me to build winlogbeat container on Linux and successfully run `winlogbeat setup`.

```Dockerfile
FROM golang:1.13
ARG version=7.3.2
ARG mage=1.9.0

# Install mage. Required for beats build processes.
RUN wget -qO- https://github.com/magefile/mage/releases/download/v${mage}/mage_${mage}_Linux-64bit.tar.gz | tar -C /bin -zxO mage > /bin/mage
RUN chmod 755 /bin/mage

# Install virtualenv. Required for beats build processes.
RUN apt-get update
# python 2's pip
RUN apt-get install -y python-pip
RUN pip install virtualenv

#RUN go get github.com/elastic/beats
RUN mkdir -p $GOPATH/src/github.com/elastic
RUN git -C $GOPATH/src/github.com/elastic clone --depth 1 --single-branch --branch v${version} https://github.com/elastic/beats
WORKDIR $GOPATH/src/github.com/elastic/beats

# Target winlogbeat
WORKDIR $GOPATH/src/github.com/elastic/beats/winlogbeat

# Patch out the `init` function in winlogbeat to avoid this error on non-Windows systems:
#   Exiting: Failed to create new event log. No event log API is available on this system
#
# This allows us to run `winlogbeat setup` on linux systems.
RUN sed -i -e 's/^\(func .* \)\(init(.*) error {.*\)$/\1\2 return nil }\n\1_\2/' beater/winlogbeat.go

# Compile winlogbeat
RUN make

# Generate winlogbeat's dashboards/templates/fields/etc
RUN make update
```

```
% docker run -it -v $PWD/winlogbeat.yml:/winlogbeat.yml:Z winlogbeat-on-linux bash -c 'install -o root /winlogbeat.yml /tmp/winlogbeat.yml; ./winlogbeat setup -c /tmp/winlogbeat.yml -E "setup.dashboards.directory=./build/kibana"'
Index setup finished.
Loading dashboards (Kibana must be running and reachable)
Loaded dashboards
```

**Comment by andrewkroh** [comment]
@jordansissel `mage` should be installed from the beat's vendor/ directory to ensure you have the same version. You can do this by running `make mage` after cloning out beats.

**Comment by jordansissel** [comment]
@andrewkroh I saw that, but it doesn't work:

```
% make update
Installing mage v1.8.0 from vendor dir.
bash: mage: command not found
../dev-tools/make/mage.mk:9: recipe for target 'mage' failed
make: [mage] Error 127 (ignored)
mage update
bash: mage: command not found
Makefile:12: recipe for target 'update' failed
make: *** [update] Error 127

% make mage
Installing mage v1.8.0 from vendor dir.
bash: mage: command not found
../dev-tools/make/mage.mk:9: recipe for target 'mage' failed
make: [mage] Error 127 (ignored)
```

**Comment by andrewkroh** [comment]
Is `GOPATH` not included in the `PATH` already? I would have thought that this container did that by default.

**Comment by jordansissel** [comment]
I think `make mage` assumes $GOPATH/bin (where `go install` sends things?) is in the $PATH

**Comment by jordansissel** [comment]
Dunno about the container, I experienced `make mage` fail on my workstation, so I manually install mage and repeated this behavior for the container.

**Comment by nedsir** [comment]
Thanks a lot @jordansissel 

I had the exact same issue and it works like a charm.

**Comment by hinchliff** [comment]
thanks @jordansissel !

here is an updated version for winlogbeat 7.16.2 -- using go 1.17.5, python 3, and the mage version directly from winlogbeat (1.11.0)

```Dockerfile
# https://github.com/elastic/beats/issues/13707
# docker build . -t winlogbeat-on-linux:7.16.2

FROM golang:1.17.5

ARG version=7.16.2

# Install virtualenv. Required for beats build processes.
RUN apt-get update --no-install-recommends

# Python 3 venv (virtualenv)
RUN apt-get install --no-install-recommends --yes python3-venv

#RUN go get github.com/elastic/beats
RUN mkdir -p $GOPATH/src/github.com/elastic
RUN git -C $GOPATH/src/github.com/elastic clone --depth 1 --single-branch --branch v${version} https://github.com/elastic/beats
WORKDIR $GOPATH/src/github.com/elastic/beats

# Target winlogbeat
WORKDIR $GOPATH/src/github.com/elastic/beats/winlogbeat

# Patch out the `init` function in winlogbeat to avoid this error on non-Windows systems:
#   Exiting: Failed to create new event log. No event log API is available on this system
#
# This allows us to run `winlogbeat setup` on linux systems.
RUN sed -i -e 's/^\(func .* \)\(init(.*) error {.*\)$/\1\2 return nil }\n\1_\2/' beater/winlogbeat.go

# Compile winlogbeat
RUN make

# Generate winlogbeat's dashboards/templates/fields/etc
RUN make update
```

this is really the only thing I needed out of mine:
```shell
docker run --rm -t winlogbeat-on-linux:7.16.2 ./winlogbeat export template --es.version 7.16.2 -E setup.template.type=legacy
```

**Comment by legoguy1000** [comment]
We've been using this to great success though now we're on 8.5.1 and it doesn't load the Ingest Pipelines.  It still loads the datastream/template/dashboards, just not the pipelines and can't figure out why.

**Comment by legoguy1000** [comment]
I did that and then found that the pipelines are loaded via the `init()` function, https://github.com/elastic/beats/blob/main/winlogbeat/beater/winlogbeat.go#L107.  So renaming the fix action breaks that.  So what I did was just copy the entire `winlogbeat.go` and removed lines 92-106 which is really the issue on linux and we'll just have to manually update this every time we update Elastic to make sure the file hasn't changed.
